### PR TITLE
Feature/il 2026

### PR DESCRIPTION
**Pull request by alvaro-ipor**
https://ipor-labs.atlassian.net/browse/IL-2026?atlOrigin=eyJpIjoiNTVmOGJiODY3YzhlNDRlZWI0YTJlM2UwOThhOTFhZGEiLCJwIjoiaiJ9

### CODE REVIEW

File: test/ivToken/IvToken.t.sol [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+contract IvTokenTest is Test, TestCommons {
+    IvToken internal _ivToken;
+    DaiMockedToken internal _daiMockedToken;
+    address internal _admin;
+    address internal _userOne;
+    address internal _userTwo;
+
+    function setUp() public {
+        _ivToken = new IvToken("IvToken", "IVT", address(0x6B175474E89094C44Da98b954EedeAC495271d0F));
```
Comment (pete-ipor) [inline test/ivToken/IvToken.t.sol:18]: Please add a description to this address `0x6B175474E89094C44Da98b954EedeAC495271d0F` (random or so)

File: test/ivToken/IvToken.t.sol [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    function testShouldNotConfirmTransferOwnershipTwiceWhenSenderNotAppointedOwner() public {
+        // given
+        IvToken ivTokenDai = new IvToken("IV DAI", "ivDAI", address(_daiMockedToken));
+        address ownerBefore = ivTokenDai.owner();
+        ivTokenDai.transferOwnership(_userOne);
+        vm.prank(_userOne);
+        ivTokenDai.confirmTransferOwnership();
+        // when
+        vm.expectRevert(abi.encodePacked("Ownable: caller is not the owner"));
+        ivTokenDai.transferOwnership(_userOne);
```
Comment (pete-ipor) [inline test/ivToken/IvToken.t.sol:103]: According to the test name, it should be something like this 
```
vm.prank(_userOne);
 ivTokenDai.confirmTransferOwnership();
```

File: test/ivToken/IvToken.t.sol [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        IvToken ivTokenDai = new IvToken("IV DAI", "ivDAI", address(_daiMockedToken));
+        address ownerBefore = ivTokenDai.owner();
+        // when 
+        ivTokenDai.transferOwnership(_userOne);
+        vm.prank(_userOne);
+        ivTokenDai.confirmTransferOwnership();
+        // then
+        vm.prank(_admin);
+        vm.expectRevert(abi.encodePacked("Ownable: caller is not the owner"));
+        ivTokenDai.transferOwnership(_userOne);
```
Comment (pete-ipor) [inline test/ivToken/IvToken.t.sol:121]: This line should be in when a section 
lines 111-117  -> // given
lines 119-121 -> // when

File: test/ivToken/IvToken.t.sol [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        assertEq(ownerBefore, _admin);
+        assertEq(ownerAfter, _userOne);
+    }
+
+    function testShouldHaveRightsToTransferOwnershipWhenSenderStillHasRights() public {
+        // given
+        IvToken ivTokenDai = new IvToken("IV DAI", "ivDAI", address(_daiMockedToken));
+        ivTokenDai.transferOwnership(_userOne);
+        // when
+        ivTokenDai.transferOwnership(_userOne);
```
Comment (pete-ipor) [inline test/ivToken/IvToken.t.sol:132]: pleas change _userOne to _userTwo

File: test/ivToken/IvTokenBurn.t.sol [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+	function testShouldNotBurnIvTokenWhenZeroAddress() public {
+		// given
+		address mockIporVaultAddress = _userOne;
+		_ivToken.setStanley(mockIporVaultAddress);
+		// when
+		vm.prank(_userOne);
+		// then
+		vm.expectRevert(abi.encodePacked("ERC20: burn from the zero address"));
+		_ivToken.burn(address(0), 1000000000000000000000000);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:51]: This should be in //when a section

File: test/ivToken/IvTokenBurn.t.sol [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+	function testShouldNotBurnWhenAmountExceedsBalance() public {
+		// given
+		address mockIporVaultAddress = _userOne;
+		_ivToken.setStanley(mockIporVaultAddress);
+		// when
+		vm.prank(_userOne);
+		// then
+		vm.expectRevert(abi.encodePacked("ERC20: burn amount exceeds balance"));
+		_ivToken.burn(mockIporVaultAddress, 1000000000000000000000001);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:62]: This should be in //when a section

Comment (pete-ipor) [reply test/ivToken/IvTokenBurn.t.sol:62]: Please mint some ivToken for this address before run burn

File: test/ivToken/IvTokenBurn.t.sol [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	function testShouldBurnTokens() public {
+		// given
+		address mockIporVaultAddress = _userOne;
+		_ivToken.setStanley(mockIporVaultAddress);
+		uint256 amount = 1000000000000000000000000;
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:69]: Please use 1e18 notation for numbers

File: test/ivToken/IvTokenBurn.t.sol [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		vm.prank(_userOne);
+		_ivToken.mint(_userOne, 1000000000000000000000000);
+		//when
+		// then
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:73]: In//when section should be `burn ` action

Comment (pete-ipor) [reply test/ivToken/IvTokenBurn.t.sol:73]: check `_userOne ` balanceOf ivTokens before and after burn

File: test/ivToken/IvTokenBurn.t.sol [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+		vm.prank(_userOne);
+		vm.expectEmit(true, true, false, true);
+		emit Transfer(_userOne, address(0), amount);
+		vm.expectEmit(true, false, false, true);
+		emit Burn(_userOne, amount);
+		_ivToken.burn(_userOne, amount);
+	}
+
+	function testShouldEmitEvent() public {
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:82]: which events

File: test/ivToken/IvTokenBurn.t.sol [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+		// given
+		address mockIporVaultAddress = _admin;
+		_ivToken.setStanley(mockIporVaultAddress);
+		uint256 amount = 1000000000000000000000000;
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:86]: Please use 1e18 notation for numbers

File: test/ivToken/IvTokenBurn.t.sol [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+		// when
+		// then
+		_ivToken.mint(_userOne, amount);
+		vm.expectEmit(true, false, false, true);
+		emit Burn(_userOne, amount);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:91]: I think this should be in //when a section (all)

File: test/ivToken/IvTokenMint.t.sol [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        _admin = address(this);
+        _userOne = _getUserAddress(1);
+        _userTwo = _getUserAddress(2);
+    }
+
+	function testShouldNotMintIvTokenWhenNotStanley () public {
+		// given
+		// when
+		// then
+		vm.expectRevert(abi.encodePacked("IPOR_501"));
```
Comment (pete-ipor) [inline test/ivToken/IvTokenMint.t.sol:28]: I think this should be in //when a section (all)

File: test/ivToken/IvTokenMint.t.sol [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+		_ivToken.mint(_userOne, 1000000000000000000000000);
+	}
+
+	function testShouldNotMintIvTokenWhenAmountIsZero() public {
+		// given
+		address mockIporVaultAddress = _userOne;
+		_ivToken.setStanley(mockIporVaultAddress);
+		// when
+		vm.prank(_userOne);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenMint.t.sol:37]: this is only setup sender,

File: test/ivToken/IvTokenMint.t.sol [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+		// then
+		vm.expectRevert(abi.encodePacked("IPOR_503"));
+		_ivToken.mint(mockIporVaultAddress, 0);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenMint.t.sol:40]: this should be in //when the section

File: test/ivToken/IvTokenMint.t.sol [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+	function testShouldNotMintIvTokenWhenZeroAddress() public {
+		// given
+		address mockIporVaultAddress = _userOne;
+		_ivToken.setStanley(mockIporVaultAddress);
+		// when
+		vm.prank(_userOne);
+		// then
+		vm.expectRevert(abi.encodePacked("ERC20: mint to the zero address"));
+		_ivToken.mint(address(0), 1000000000000000000000000);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenMint.t.sol:51]: This should be in //when section and use 1e18 notation

File: test/ivToken/IvTokenMint.t.sol [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+		_ivToken.setStanley(mockIporVaultAddress);
+		uint256 amount = 1000000000000000000000000;
+		// when
+		vm.prank(_userOne);
+		// then
+		vm.expectEmit(true, true, false, true);
+		emit Transfer(address(0), _userOne, amount);
+		vm.expectEmit(true, false, false, true);
+		emit Mint(_userOne, amount);
+		_ivToken.mint(_userOne, amount);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenMint.t.sol:66]: This should be in //when section and use 1e18 notation

File: test/ivToken/IvTokenMint.t.sol [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	}
```
Comment (pete-ipor) [inline test/ivToken/IvTokenMint.t.sol:67]: Check `_userOne ` balance of ivToken before and after mint

**Review by pete-ipor — CHANGES_REQUESTED** [review]

File: test/ivToken/IvTokenMint.t.sol (hunk 13 above)
Comment (alvaro-ipor) [reply test/ivToken/IvTokenMint.t.sol:37]: I set up the sender to get `IPOR_503`. Otherwise I get `IPOR_501`

Comment (pete-ipor) [reply test/ivToken/IvTokenMint.t.sol:36]: I know, but this is not the main action in the test. In the `when` section, there should be `mint` token

Comment (alvaro-ipor) [reply test/ivToken/IvTokenMint.t.sol:36]: Do you mean this?

```
	function testShouldNotMintIvTokenWhenAmountIsZero() public {
		// given
		address mockIporVaultAddress = _admin;
		_ivToken.setStanley(mockIporVaultAddress);
		// when
		_ivToken.mint(_userOne, 1*10**18);
		vm.expectRevert(abi.encodePacked("IPOR_503"));
		_ivToken.mint(_userOne, 0);
	}
```

Comment (pete-ipor) [reply test/ivToken/IvTokenMint.t.sol]: i would say 
```
	function testShouldNotMintIvTokenWhenAmountIsZero() public {
		// given
		address mockIporVaultAddress = _admin;
		_ivToken.setStanley(mockIporVaultAddress);
		// when
		vm.expectRevert(abi.encodePacked("IPOR_503"));
		_ivToken.mint(_userOne, 0);
	}
```

File: test/ivToken/IvToken.t.sol [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+contract IvTokenTest is Test, TestCommons {
+    IvToken internal _ivToken;
+    MockTestnetTokenDai internal _mockTestnetTokenDai;
+    address internal _admin;
+    address internal _userOne;
+    address internal _userTwo;
+
+    function setUp() public {
+        _ivToken = new IvToken("IvToken", "IVT", address(0x6B175474E89094C44Da98b954EedeAC495271d0F)); // random address
+        _mockTestnetTokenDai = new MockTestnetTokenDai(1*10**18);
```
Comment (pete-ipor) [inline test/ivToken/IvToken.t.sol:18]: instead of `1*10**18` you can use Constants.D18

File: test/ivToken/IvTokenBurn.t.sol [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+        _userOne = _getUserAddress(1);
+        _userTwo = _getUserAddress(2);
+    }
+
+	function testShouldNotBurnIvTokenWhenNotStanley () public {
+		// given
+		// when
+		// then
+		vm.expectRevert(abi.encodePacked("IPOR_501"));
+		_ivToken.burn(_userOne, 1*10**18);
```
Comment (pete-ipor) [inline test/ivToken/IvTokenBurn.t.sol:29]: Move it to `when` section

File: test/ivToken/IvTokenMint.t.sol (hunk 17 above)
Comment (pete-ipor) [reply test/ivToken/IvTokenMint.t.sol:63]: ???

**Review by pete-ipor — APPROVED** [review]

**Comment by pete-ipor** [comment]
Now you can remove tests in hardhat and marge branch

**Review by mario-ipor — APPROVED** [review]

**Pull request merged**